File: frontend-web/src/modules/convenience/services/productTourService.js
// src/modules/convenience/services/productTourService.js
// src/modules/convenience/services/productTourService.js

import api from "../../../api/axios";

const API_URL = "/convenience/product-tours";

export const productTourService = {
  // =====================================================
  // GET CONFIGURABLE
  // =====================================================

  // GET /api/convenience/product-tours/pending-config
  getPendingConfig: async () => {
    const response = await api.get(`${API_URL}/pending-config`);

    return response.data;
  },

  // =====================================================
  // POST CONFIG
  // =====================================================

  // POST /api/convenience/product-tours/{assignmentId}/config
  configure: async (assignmentId, data) => {
    const response = await api.post(`${API_URL}/${assignmentId}/config`, data);

    return response.data;
  },

  // =====================================================
  // PATCH CONFIG
  // =====================================================

  // PATCH /api/convenience/product-tours/{assignmentId}/config
  updateConfig: async (assignmentId, data) => {
    const response = await api.patch(`${API_URL}/${assignmentId}/config`, data);

    return response.data;
  },
  // =====================================================
  // GET ALL
  // =====================================================

  // GET /api/convenience/product-tours
  getAll: async () => {
    const response = await api.get(API_URL);

    return response.data;
  },

  // =====================================================
  // GET CONFIGURATION HISTORY
  // =====================================================

  // GET /api/convenience/product-tours/configuration-history
  getConfigurationHistory: async () => {
    const response = await api.get(`${API_URL}/configuration-history`);

    return response.data;
  },

  // =====================================================
  // GET CONFIGURATION HISTORY DETAIL
  // =====================================================

  // GET /api/convenience/product-tours/tour/{tourId}
  getConfigurationHistoryDetail: async (tourId) => {
    const response = await api.get(`${API_URL}/tour/${tourId}`);

    return response.data;
  },

  // =====================================================
  // COMPLETE CONFIGURATION
  // =====================================================

  // POST /api/convenience/product-tours/{tourId}/complete
  completeConfiguration: async (tourId) => {
    const response = await api.post(`${API_URL}/${tourId}/complete`);

    return response.data;
  },
};

export default productTourService;

File: backend/convenience-service/src/main/java/com/project/convenience/service/product/ProductTourService.java
package com.project.convenience.service.product;

import com.project.convenience.dto.product.convenience.HistoryProductTourResponse;
import com.project.convenience.dto.product.convenience.ProductTourResponse;
import com.project.convenience.mapper.HistoryProductTourMapper;
import com.project.convenience.mapper.ProductTourMapper;
import com.project.convenience.model.ProductTour;
import com.project.convenience.model.enums.ProductTourStatus;
import com.project.convenience.repository.HistoryProductTourRepository;
import com.project.convenience.repository.ProductTourRepository;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@Transactional
public class ProductTourService {

        private final ProductTourRepository productTourRepository;
        private final ProductTourMapper mapper;
        private final HistoryProductTourRepository historyProductTourRepository;

        public ProductTourService(
                        ProductTourRepository productTourRepository,
                        ProductTourMapper mapper,
                        HistoryProductTourRepository historyProductTourRepository) {

                this.productTourRepository = productTourRepository;
                this.mapper = mapper;
                this.historyProductTourRepository = historyProductTourRepository;
        }

        public void createProductTourFromEvent(
                        UUID tourId,
                        UUID cruiseAreaId) {

                boolean exists = productTourRepository
                                .findByTourIdAndCruiseAreaId(
                                                tourId,
                                                cruiseAreaId)
                                .isPresent();

                if (exists) {
                        return;
                }

                ProductTour productTour = new ProductTour();

                productTour.setTourId(tourId);
                productTour.setCruiseAreaId(cruiseAreaId);
                productTour.setStatus(
                                ProductTourStatus.WAITING_CONFIG);

                productTourRepository.save(productTour);
        }

        // =========================================================
        // GET ALL PRODUCT TOURS
        // =========================================================

        @Transactional(readOnly = true)
        public List<ProductTourResponse> getAllAssignments() {

                return productTourRepository
                                .findAll()
                                .stream()
                                .map(mapper::toProductTourResponse)
                                .toList();
        }

        // =========================================================
        // GET PRODUCT TOUR CẦN CẤU HÌNH
        // =========================================================

        @Transactional(readOnly = true)
        public List<ProductTourResponse> getPendingConfig() {

                return productTourRepository
                                .findConfigurable(
                                                List.of(
                                                                ProductTourStatus.WAITING_CONFIG))
                                .stream()
                                .map(mapper::toProductTourResponse)
                                .toList();
        }
        // =========================================================
        // GET CONFIGURATION HISTORY
        // =========================================================

        @Transactional(readOnly = true)
        public List<HistoryProductTourResponse> getConfigurationHistory() {

                return historyProductTourRepository
                                .findAllByOrderByCompletedAtDesc()
                                .stream()
                                .map(HistoryProductTourMapper::toResponse)
                                .toList();
        }
        // =========================================================
        // GET CONFIGURATION HISTORY DETAIL
        // =========================================================

        @Transactional(readOnly = true)
        public List<ProductTourResponse> getConfigurationHistoryDetail(
                        UUID tourId) {

                return productTourRepository
                                .findAllByTourIdOrderByCreatedAtAsc(tourId)
                                .stream()
                                .map(mapper::toProductTourResponse)
                                .toList();
        }
}